### DISCUSSION TITLE
damage + elastic occur error

### DISCUSSION THREAD

**Discussion by IntePNum** [peridigm/peridigm#333] [General]
My model is plate with a rect-hole in the center.

My material model is "Elastic", Damage model is "Critical Stretch"

I applied two solvers, Solver1 for deformation, Solver2 for damage.

When it calculates to damage, it would appear "un convergence".

why??
who can help me. I would give a great thanks. Please!

This is my yaml.file
*******************************************************************************************
Peridigm:

#     Description
#       Simulation of an elastic tensile bar subjected to 0.5% strain.
#       Engineering strain is estimated by tracking the elongation of a
#       segment of the bar using the Nearest_Point_Data compute class.
#       Engineering stress is estimated by tracking the reaction forces
#       in the end segments of the bar using the Block_Data compute class.
#       Using the data computed by the compute classes, engineering strain
#       can be determined as the change in length divided by the initial
#       length, and engineering stress as the force in an end segment
#       divided by the initial cross-sectional area.
#
#     Units are CGS
#       mass          g
#       length        cm
#       time          s
#       density       g/cm^3
#       force         dyne
#       pressure      dyne/cm^2
#       velocity      cm/s
#       acceleration  cm/s^2
#       energy        erg

  Verbose: false
  Discretization:
      Type: "Text File"
      Input Mesh File: "allnode_set.txt"

  Materials:
      My Material:
        Material Model: "Elastic"
        Density: 8.0
        Young's Modulus: 192.0e9
        Poisson's Ratio: 0.33
               

        
  Damage Models:
      My Damage Model:
        Damage Model: "Critical Stretch"
        Critical Stretch: 0.02

  Blocks:
      Block One:
        Block Names: "block_1"
        Material: "My Material"
        Horizon: 0.75e-3
      Block Two:
        Block Names: "block_2"
        Material: "My Material"
        Damage Model: "My Damage Model"
        Horizon: 0.75e-3
      Block Three:
        Block Names: "block_3"
        Material: "My Material"
        Horizon: 0.75e-3

  Boundary Conditions:
      Node Set One: "node_set_1.txt"
      Node Set Two: "node_set_2.txt"
      Node Set Three: "node_set_3.txt"
      Prescribed Bottom Displacement X:
        Type: "Prescribed Displacement"
        Node Set: "Node Set One"
        Coordinate: "x"
        Value: "0" 
      Prescribed Bottom Displacement Y:
        Type: "Prescribed Displacement"
        Node Set: "Node Set One"
        Coordinate: "y"
        Value: "-2.7541e-7*t"  
      Prescribed Bottom Displacement Z:
        Type: "Prescribed Displacement"
        Node Set: "Node Set One"
        Coordinate: "z"
        Value: "0"  
      Prescribed Uppon Displacement Y:
        Type: "Prescribed Displacement"
        Node Set: "Node Set Three"
        Coordinate: "y"
        Value: "2.7541e-7*t"  
      Prescribed Uppon Displacement X:
        Type: "Prescribed Displacement"
        Node Set: "Node Set Three"
        Coordinate: "x"
        Value: "0"  
      Prescribed Uppon Displacement Z:
        Type: "Prescribed Displacement"
        Node Set: "Node Set Three"
        Coordinate: "z"
        Value: "0"  
        
          
  Solver1:
      Verbose: false
      Initial Time: 0.0
      Final Time: 750.0
      QuasiStatic:
        Number of Load Steps: 750
        Absolute Tolerance: 1.0
        Maximum Solver Iterations: 20
  Solver2:
      Verbose: false
      Initial Time: 750.0
      Final Time: 1500.0
      QuasiStatic:
        Number of Load Steps: 750000
        Absolute Tolerance: 1.0
        Maximum Solver Iterations: 20

  Output:
      Output File Type: "ExodusII"
      Output Filename: "test"
      Output Frequency: 1000
      Output Variables:
        Displacement: true
        Velocity: true
        Element_Id: true
        Proc_Num: true
        Force_Density: true
        Force: true
        Damage: true
        Volume: true
        Radius: true
        Number_Of_Neighbors: true
        Neighborhood_Volume: true
![2](https://github.com/peridigm/peridigm/assets/108223279/ed138959-e48b-4e43-885a-64e2b7ed2ded)

*******************************************************************************************
![1](https://github.com/peridigm/peridigm/assets/108223279/259b9c12-9d25-458f-ada6-d5bdb3dd8cc5)

**Comment by DP52** [comment]
Sorry I don 't know your problem. But I want to ask another question, how to get its stress using elastic materials ?

**Reply by IntePNum** [reply]
I am sorry, I don't know either......I contact this method for one month.....ahahaha.sorry!!!!

**Comment by andrisfreimanis** [comment]
If I remember this correctly, Elastic is an ordinary state based PD material that computes force density (N/m3) not stress (N/m2). You could calculate engineering stress by multiplying the stress density over some layer of nodes with the height of that layer. If you want Peridigm to calculate stress, you can use a non-ordinary state-based PD model Elastic Correspondence (check spelling) and that will calculate stress.

Best regards
Andris
________________________________

**Reply by DP52** [reply]
Thank you for your reply, but with the Elastic Correspondence material model and the critical stretch damage model, crack propagation will cause cracks to just open and not expand. 
And I've tried other damage models, none of which are currently available, so there's no better way to do it than modify the program? Thanks again for your reply!

Best regards
Alin

**Comment by andrisfreimanis** [comment]
You could post the error message, otherwise it's hard to know what's going on.

Best regards
Andris
________________________________

**Reply by IntePNum** [reply]
![image](https://github.com/peridigm/peridigm/assets/108223279/6d62d2d0-a417-4b25-bbe4-ba9cc9e8f15c)
thanks for your reply!
The error is uncoverged,such as show in this picture.
Best wishes,
Intemp

**Comment by andrisfreimanis** [comment]
It's not an error, the solver just doesn't converge. You can try a smaller timestep, adaptive time-stepping (check tests for examples), or the explicit solver. When the damage starts, implicit solver can have troubles converging.

Andris
________________________________

**Reply by IntePNum** [reply]
ok, prof,

Is the explicit solver suiable for quasi-static problem?

![image](https://github.com/peridigm/peridigm/assets/108223279/45cc8e6a-e928-413d-8e69-9fdeec4aab0e)

**Comment by djlittl** [comment]
That’s the Velocity Verlet time integrator, which is explicit dynamics.  That formulation does not assume static equilibrium, hence the simulation will not be a quasi-static simulation.  It will do much much better at progressing through damage (which is very likely not a quasi-static process anyway).  The downside is that many small time steps will be required.  Depending on the mesh size and material stiffness, this may or may not be tractable.

To answer your question directly, yes, that seems reasonable.  If you run into instability problems, try reducing the safety factor to maybe 0.5.

**Reply by IntePNum** [reply]
hey prof,

If I want to simulate a quasi-static problem with damage,

what is the best choice for the constitutive model and damage model?

thanks for your reply.

best wishes!
intep
